packages/fast-element/src/observation/subscriber-collection.ts

```diff
@@ -0,0 +1,105 @@
+/**
+ * Implemented by objects that are interested in change notifications.
+ */
+export interface ISubscriber {
+    handleChange(source: any, args: any): void;
+}
+
+/**
+ * Efficiently keeps track of subscribers interested in change notifications.
+ *
+ * @remarks
+ * This collection is optimized for the most common scenario of 1 - 3 subscribers.
+ * With this in mind, it can store a subscriber in an internal field, allowing it to avoid Array#push operations.
+ * If the collection ever exceeds three subscribers, it will spill over into an array.
+ */
+export class SubscriberCollection {
+    private sub1: ISubscriber | undefined = void 0;
+    private sub2: ISubscriber | undefined = void 0;
+    private sub3: ISubscriber | undefined = void 0;
+    private spillover: ISubscriber[] | undefined = void 0;
+
+    public hasSubscriber(subscriber: ISubscriber) {
+        if (this.sub1 === subscriber) return true;
+        if (this.sub2 === subscriber) return true;
+        if (this.sub3 === subscriber) return true;
+        return this.spillover !== void 0 && this.spillover.indexOf(subscriber) !== -1;
+    }
+
+    public addSubscriber(subscriber: ISubscriber) {
+        if (this.hasSubscriber(subscriber)) {
+            return;
+        }
+
+        if (this.sub1 === void 0) {
+            this.sub1 = subscriber;
+            return;
+        }
+
+        if (this.sub2 === void 0) {
+            this.sub2 = subscriber;
+            return;
+        }
+
+        if (this.sub3 === void 0) {
+            this.sub3 = subscriber;
+            return;
+        }
+
+        if (this.spillover === void 0) {
+            this.spillover = [];
+        }
+
+        this.spillover.push(subscriber);
+    }
+
+    public removeSubscriber(subscriber: ISubscriber) {
+        if (this.sub1 === subscriber) {
+            this.sub1 = void 0;
+            return;
+        }
+
+        if (this.sub2 === subscriber) {
+            this.sub2 = void 0;
+            return;
+        }
+
+        if (this.sub3 === subscriber) {
+            this.sub3 = void 0;
+            return;
+        }
+
+        if (this.spillover !== void 0) {
+            const index = this.spillover.indexOf(subscriber);
```

Comment:
Should this *also* set `spillover` to undefined if the there are no more items? It seems you could avoid re-entering both this if statement and the statement on line 99 un-necessarily, though I'm not sure how much that really matters.

Reply:
I think we'd have to do some specific benchmarking around that to determine if it would make a big difference or not. I'd prefer to leave it as is for now.